### DISCUSSION TITLE
Can I use the same pgxmock.NewPool() across tests?

### DISCUSSION THREAD

**Discussion by aravindhp** [pashagolub/pgxmock#214] [Q&A] [answered]
I am testing an API server by doing the following:

```go
func TestServer(t *testing.T) {

	// Get the swagger description of our API
	swagger, err := api.GetSwagger()
	require.NoError(t, err)

	// Create a new ServeMux for testing.
	m := http.NewServeMux()
	opts := api.StdHTTPServerOptions{
		BaseRouter: m,
	}

	mockPool, err := pgxmock.NewPool()
	require.NoError(t, err)
	defer mockPool.Close()

	mockdb := db.New(mockPool)
	api.HandlerWithOptions(mockdb, opts)

	t.Run("ABC", func(t *testing.T) {
		mockPool.ExpectQuery(abcQuery).
			WithArgs(pgxmock.AnyArg()).
			WillReturnRows(abcRow)
		rr := testutil.NewRequest().Post("/abc").GoWithHTTPHandler(t, m).Recorder
		assert.Equal(t, http.StatusAccepted, rr.Code)

	})

	t.Run("XYZ", func(t *testing.T) {
		mockPool.ExpectQuery(xyzQuery).
			WithArgs(pgxmock.AnyArg()).
			WillReturnRows(xyzRow)
		rr := testutil.NewRequest().Post("/xyz").GoWithHTTPHandler(t, m).Recorder
		assert.Equal(t, http.StatusAccepted, rr.Code)
	})
}
```

For simplicity I have left out most of the code. What I am seeing is that if I run the tests individually they pass but if I run `TestServer()` one of the tests fail with `Error getting action: could not match actual sql` where the mismatch is happening across tests. I.E. test `XYZ` fails because the actual SQL query has a mismatch the regex specified in `abcQuery`. If use separate a local `mockPool` in each test I don't see this issue. Am I following a bad pattern here? Should I be instantiating a new mock pool for each test?

**Comment by pashagolub** [comment, marked as answer]
Hello.

>  Am I following a bad pattern here?

By default pgxmock expects commands to be executed in order. If you run tests in parallel, of course, there is no opportunity to keep that order. It is possible to explicitly ignore the order of the commands:

```go
	mock, _ := NewConn()
	mock.MatchExpectationsInOrder(false)
...
// run your magic
```

Is it a bad pattern to use one global mock? Not really, sometimes (not that often though) you need to stress your code in parallel and that might be a good idea.

>Should I be instantiating a new mock pool for each test?

I prefer to use local `mockPool` for every test:
- it's lightweight and doesn't stress the system;
- it's easier to control expectations and flows;
- it doesn't interfere with other tests and bugs can be localized easier.

Best regards
